prog/redeliver_github_failures.rb

```diff
@@ -3,14 +3,23 @@
 class Prog::RedeliverGithubFailures < Prog::Base
   label def wait
     last_check_time = Time.parse(frame["last_check_at"])
-    failed = failed_deliveries(last_check_time).each do |delivery|
-      Clog.emit("redelivering failed delivery") { {delivery: delivery.to_h} }
-      client.post("/app/hook/deliveries/#{delivery[:id]}/attempts")
-    end.count
-    Clog.emit("redelivered failed deliveries") { {deliveries: {failed:}} }
-    update_stack({"last_check_at" => Time.now})
+    nap 10 if Time.now - last_check_time < 2 * 60
+    failures = failed_deliveries(last_check_time)
+    failures.each_slice(50) do |deliveries|
+      bud Prog::RedeliverGithubFailures, {"delivery_ids" => deliveries.map { it[:id] }}, "redeliver"
+    end
+    update_stack({"last_check_at" => Time.now.to_s})
+    hop_wait_redelivers
+  end
+
+  label def wait_redelivers
+    register_deadline("wait", 10 * 60)
+    reap(:wait)
+  end
 
-    nap 2 * 60
+  label def redeliver
+    frame["delivery_ids"].each { client.post("/app/hook/deliveries/#{it}/attempts") }
```

Comment:
With slower than expected GitHub response times, I think this could still lead to apoptosis, though it would require an average of ~2.4s response time per request. That seems unlikely enough that we probably don't need to worry about it. In the event it does become an issue, we could start checking each iteration how much lease time is remaining, and if it isn't sufficient, modify `frame["delivery_ids"]` and `nap 0`.

Reply:
Since we've set the timeout to 5 seconds, do you think redelivering in batches of 25 instead of 50 would be a safer approach?

https://github.com/ubicloud/ubicloud/blob/3bdf60d49f62defc266c18695ad5d5ed2b584fa8/lib/github.rb#L11

Reply:
It's a safer approach for avoiding apoptosis. The tradeoff is it doubles the number of strands, which makes it more likely tie up the first respirate worker. However, if we change `st` ubids to be fully random and not timestamp based (it least for certain `st` ubids if not for all), then I agree that 25 is better.

@pykello @fdr What are your thoughts on making `st` ubids be random instead of timestamp based? When originally designed, I don't think this was an issue, and using timestamp ubids provided basically free storage for the created at timestamp, but since the switch to partitioned respirate, it has become more of an issue. It doesn't just affect this prog type, but all `st` strands.